File: Diagrammatize.Design.Pattern/code/cxx/Composite/Sample/Directory.cpp
#include "Directory.hpp"

#include <iostream>

Directory::Directory(const std::string& name_): name(name_) {
}

std::string Directory::getName() const {
    return name;
}

int Directory::getSize() const {
    int size = 0;
    for (auto entry: directory) {
        size += entry->getSize();
    }
    return size;
}

Entry* Directory::add(Entry* entry) {
    directory.push_back(entry);
    return this;
}

void Directory::printList(const std::string& prefix) const {
    std::cout << prefix << "/" << *this << std::endl;
    for (auto entry: directory) {
        entry->printList(prefix + "/" + name);
    }
}

File: Diagrammatize.Design.Pattern/code/python/Composite/Sample/Directory.py
from Entry import Entry

class Directory(Entry):
    def __init__(self, name):
        self.name = name
        self.directory = list()

    def getName(self):
        return self.name

    def getSize(self):
        size = 0
        for entry in self.directory:
            size += entry.getSize()
        return size

    def add(self, entry):
        self.directory.append(entry)
        return self

    def printListAux(self, prefix):
        print(prefix + "/" + str(self))
        for entry in self.directory:
            entry.printListAux(prefix + "/" + self.name)
